bench/Libraries/Microsoft.Extensions.Http.Diagnostics.PerformanceTests/Benchmarks/HeadersReaderBenchmark.cs

```diff
@@ -0,0 +1,107 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Net.Http;
+using BenchmarkDotNet.Attributes;
+using Microsoft.Extensions.Compliance.Testing;
+using Microsoft.Extensions.Http.Logging;
+using Microsoft.Extensions.Http.Logging.Internal;
+using Microsoft.Extensions.Telemetry.Internal;
+
+namespace Microsoft.Extensions.Http.Diagnostics.Bench.Benchmarks;
+
+public class HeadersReaderBenchmark
+{
+    private List<KeyValuePair<string, string>> _outputBuffer = null!;
+    private HttpHeadersReader _headersReader = null!;
+
+    [Params(0, 5, 15)]
+    public int HeadersCount { get; set; }
+
+    // This one can't be 0, because in that case HttpRequestReader simply doesn't call HttpHeadersReader
+    [Params(1, 3, 5, 10)]
+    public int HeadersToLogCount { get; set; }
+
+    public HttpRequestMessage? Request { get; set; }
+
+    [GlobalSetup]
+    public void Setup()
+    {
+        _outputBuffer = new(capacity: 10240);
+        Request = new HttpRequestMessage(HttpMethod.Post, "https://www.microsoft.com");
+        for (var i = 0; i < HeadersCount; i++)
+        {
+            Request.Headers.Add($"Header{i}", $"Value{i}");
+        }
+
+        var options = new LoggingOptions();
+        for (var i = 0; i < HeadersToLogCount; i++)
+        {
+            options.RequestHeadersDataClasses.Add($"Header{i}", FakeTaxonomy.PublicData);
+        }
+
+        var redactor = new HttpHeadersRedactor(ErasingRedactorProvider.Instance);
+        _headersReader = new HttpHeadersReader(new StaticOptionsMonitor<LoggingOptions>(options), redactor);
+    }
+
+    [GlobalCleanup]
+    public void Cleanup()
+    {
+        Request?.Dispose();
+        _outputBuffer.Clear();
+        _outputBuffer = null!;
+    }
+
+    [Benchmark]
+    public void HeadersReaderOld()
+    {
+        _headersReader.ReadRequestHeaders(Request!, _outputBuffer);
+    }
+
+    [Benchmark]
+    public void HeadersReaderNew()
+    {
+        _headersReader.ReadRequestHeadersNew(Request!, _outputBuffer);
+    }
+}
+
+/*
+
+BenchmarkDotNet=v0.13.5, OS=Windows 11 (10.0.22631.2861), VM=Hyper-V
```

Comment:
@geeknoid can you please take a look at these results and share your thoughts? The updated logic (`HeadersReaderNew`) works much faster when there are no headers or when headers count is less than the number of headers to log (specified via options). But in other cases it works a bit slower because it adds an overhead of getting headers count (~ 20ns).

Reply:
Hmmm. I left a comment here earlier, but where did it go?

Anyway, I think this is reasonable. Let's just mention the extra perf impact in the docs.

Reply:
Done